src/python/pants/jvm/dependency_inference/artifact_mapper.py

```diff
@@ -56,10 +65,19 @@ def addresses_for_coordinates(
 
 
 class MutableTrieNode:
+
+    __slots__ = [
```

Comment:
I've heard of `slots` but not familiar with what it does. What is the benefit of using it?

Reply:
Nit: extra newline preceding

Reply:
@tdyas In the usual case, when you declare a Python class, the interpreter creates a `dict` to store attributes and their values. Every attribute lookup requires hashing the name of the attribute and looking it up in a dict.

Declaring `__slots__` has two performance optimisations -- the first is that you save the RAM from having a dict in every instance of the `TrieNode`s. The second, and in this case most important, is that accessing the attributes of the nodes turns into an array lookup (i.e. like a C `struct`), rather than hashing and looking up into a dict.

There's no downside now, but once the mapping gets big, this will be a huge performance boost at trie construction time, and a small one at inference time.